### PR TITLE
add support for importing certificates from file

### PR DESCRIPTION
**Pull request by dagheyman**
not implemented yet:
  - overwrite confirmation prompt

### CODE REVIEW

File: debian/control [hunk 1]
```diff
@@ -23,6 +23,7 @@ Build-Depends: debhelper (>= 9),
                qml-module-io-thp-pyotherside,
                qml-module-qt-labs-settings,
                qml-module-qt-labs-platform,
+               qml-module-qt-labs-folderlistmodel,
```
Comment (emlun) [inline debian/control]: This might be overly pedantic, but let's sort these? :smile:

File: debian/control [hunk 2]
```diff
@@ -42,6 +43,7 @@ Depends: ${misc:Depends},
          qml-module-qtquick-dialogs,
          qml-module-qt-labs-settings,
          qml-module-qt-labs-platform,
+         qml-module-qt-labs-folderlistmodel,
```
Comment (emlun) [inline debian/control]: Ditto

File: doc/development.adoc [hunk 3]
```diff
@@ -43,6 +43,7 @@ Build and install the Qt5 plugin http://pyotherside.readthedocs.io/en/latest/#bu
       qml-module-qtquick-window2 \
       qml-module-qt-labs-platform \
       qml-module-qt-labs-settings \
+      qml-module-qt-labs-folderlistmodel \
```
Comment (emlun) [inline doc/development.adoc]: Ditto

File: ykman-gui/py/yubikey.py [hunk 4]
```diff
@@ -578,6 +579,59 @@ def piv_unblock_pin(self, puk, new_pin):
...
+                except (ValueError, TypeError):
+                    pass
+                try:
+                    parse_private_key(data, password=None)
+                    return {'success': True, 'error': None}
+                except (ValueError, TypeError):
+                    pass
+            raise ValueError('Failed to parse certificate or key')
+        except Exception as e:
+            return {'success': False, 'error': str(e)}
```
Comment (emlun) [inline ykman-gui/py/yubikey.py]: Suggestion: 
```
return {'success': False, 'error': type(e).__name__, 'message': str(e)}
```
to make this integrate nicely with #90. Leaving the `error` out and specifying only `message` works too.

File: ykman-gui/py/yubikey.py [hunk 5]
```diff
@@ -578,6 +579,59 @@ def piv_unblock_pin(self, puk, new_pin):
...
                 }
 
+    def piv_can_parse(self, file_url):
+        try:
+            file_path = urllib.parse.urlparse(file_url).path
+            with open(file_path, 'r+b') as file:
+                data = file.read()
+                try:
+                    parse_certificate(data, password=None)
+                    return {'success': True, 'error': None}
```
Comment (emlun) [inline ykman-gui/py/yubikey.py]: I don't think the `'error': None` is needed, is it?

File: ykman-gui/py/yubikey.py [hunk 6]
```diff
@@ -578,6 +579,59 @@ def piv_unblock_pin(self, puk, new_pin):
...
+                except (ValueError, TypeError):
+                    pass
+                try:
+                    parse_private_key(data, password=None)
+                    return {'success': True, 'error': None}
```
Comment (emlun) [inline ykman-gui/py/yubikey.py]: I don't think the `'error': None` is needed, is it?

File: ykman-gui/py/yubikey.py [hunk 7]
```diff
@@ -578,6 +579,59 @@ def piv_unblock_pin(self, puk, new_pin):
...
+                with self._open_piv() as controller:
+                    auth_failed = self._piv_ensure_authenticated(
+                        controller, pin, mgm_key)
+                    if auth_failed:
+                        return auth_failed
+                    if is_cert:
+                        controller.import_certificate(SLOT[slot], cert)
+                    if is_private_key:
+                        controller.import_key(SLOT[slot], private_key)
+            return {'success': True, 'error': None}
```
Comment (emlun) [inline ykman-gui/py/yubikey.py]: I don't think the `'error': None` is needed, is it?

File: ykman-gui/py/yubikey.py [hunk 8]
```diff
@@ -578,6 +579,59 @@ def piv_unblock_pin(self, puk, new_pin):
...
+        except Exception as e:
+            logger.error('Failed to import PIV certificate', exc_info=e)
+            return {'success': False, 'error': str(e)}
```
Comment (emlun) [inline ykman-gui/py/yubikey.py]: Suggestion: 
```
return {'success': False, 'error': type(e).__name__, 'message': str(e)}
```
to make this integrate nicely with #90. Leaving the `error` out and specifying only `message` works too.

File: ykman-gui/qml/PivCertificateInfo.qml [hunk 9]
```diff
@@ -26,6 +27,48 @@ ColumnLayout {
...
+    function importCertificate(fileUrl) {
+
+        function handleResponse(resp) {
+            if (resp.success) {
+                pivSuccessPopup.open()
+            } else {
+                if (resp.error === 'failed_parsing') {
+                    pivError.show('Something went wrong with reading the file.')
+                } else {
+                    pivError.show(resp.error)
```
Comment (emlun) [inline ykman-gui/qml/PivCertificateInfo.qml]: This might also benefit from #90.

File: ykman-gui/qml/YubiKey.qml [hunk 10]
```diff
@@ -436,4 +436,13 @@ Python {
     function pivReadCertificate(slot, cb) {
         doCall('yubikey.controller.piv_read_certificate', [slot], cb)
     }
+
+    function pivImportFile(slot, fileUrl, password, pin, managementKey, cb) {
+        doCall('yubikey.controller.piv_import_file',
```
Comment (emlun) [inline ykman-gui/qml/YubiKey.qml]: Let's use `doPivCall` here after merging (`doPivCall` is in the `piv` branch but not this one).

File: ykman-gui/qml/YubiKey.qml [hunk 11]
```diff
@@ -436,4 +436,13 @@ Python {
...
+               [slot, fileUrl, password, pin, managementKey], cb)
+    }
+
+    function pivCanParse(fileUrl, cb) {
+        doCall('yubikey.controller.piv_can_parse', [fileUrl], cb)
```
Comment (emlun) [inline ykman-gui/qml/YubiKey.qml]: Let's use `doPivCall` here after merging (`doPivCall` is in the `piv` branch but not this one).

File: debian/control (hunk 1 above)
Comment (dagheyman) [reply debian/control]: That is a good idea, but will do it in a separate PR to master.

File: ykman-gui/qml/YubiKey.qml (hunk 11 above)
Comment (dagheyman) [reply ykman-gui/qml/YubiKey.qml]: Merged piv into PR, updated calls.

**Pull request merged**